// @antv/g2plot/esm/plots/radial-bar/index.js
import { __extends } from "tslib";
import { Plot } from '../../core/plot';
import { deepAssign } from '../../utils';
import { adaptor, meta } from './adaptor';
/**
 * 玉珏图
 */
var RadialBar = /** @class */ (function (_super) {
    __extends(RadialBar, _super);
    function RadialBar() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'radial-bar';
        return _this;
    }
    /**
     * @override
     * @param data
     */
    RadialBar.prototype.changeData = function (data) {
        this.updateOption({ data: data });
        // 更新玉珏图的scale
        meta({ chart: this.chart, options: this.options });
        this.chart.changeData(data);
    };
    /**
     * 获取默认配置
     */
    RadialBar.prototype.getDefaultOptions = function () {
        return deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            interactions: [{ type: 'element-active' }],
            legend: false,
            tooltip: {
                showMarkers: false,
            },
            xAxis: {
                grid: null,
                tickLine: null,
                line: null,
            },
            maxAngle: 240,
        });
    };
    /**
     * 获取适配器
     */
    RadialBar.prototype.getSchemaAdaptor = function () {
        return adaptor;
    };
    return RadialBar;
}(Plot));
export { RadialBar };
