// @antv/g2plot/lib/plots/box/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = exports.tooltip = exports.legend = void 0;
var util_1 = require("@antv/util");
var common_1 = require("../../adaptor/common");
var utils_1 = require("../../utils");
var utils_2 = require("../../utils");
var constant_1 = require("../../constant");
var constant_2 = require("./constant");
var utils_3 = require("./utils");
/**
 * 字段
 * @param params
 */
function field(params) {
    var chart = params.chart, options = params.options;
    var xField = options.xField, yField = options.yField, groupField = options.groupField, color = options.color;
    var yFieldName = Array.isArray(yField) ? constant_2.BOX_RANGE : yField;
    var geometry = chart.schema().position(xField + "*" + yFieldName).shape('box');
    // set group field as color channel
    if (groupField) {
        geometry.color(groupField, color).adjust('dodge');
    }
    chart.data(utils_3.transformData(options.data, yField));
    return params;
}
function outliersPoint(params) {
    var chart = params.chart, options = params.options;
    var xField = options.xField, data = options.data, outliersField = options.outliersField, outliersStyle = options.outliersStyle, padding = options.padding;
    if (!outliersField)
        return params;
    var outliersView = chart.createView({ padding: padding, id: constant_2.OUTLIERS_VIEW_ID });
    outliersView.data(data);
    outliersView.axis(false);
    var geometry = outliersView.point().position(xField + "*" + outliersField).shape('circle');
    /**
     * style 的几种情况
     * g.style({ fill: 'red' });
     * g.style('x*y*color', (x, y, color) => ({ fill: 'red' }));
     */
    if (util_1.isFunction(outliersStyle)) {
        geometry.style(xField + "*" + outliersField, function (_x, _outliers) {
            var _a;
            return outliersStyle((_a = {},
                _a[xField] = _x,
                _a[outliersField] = _outliers,
                _a));
        });
    }
    else if (util_1.isObject(outliersStyle)) {
        geometry.style(outliersStyle);
    }
    return params;
}
/**
 * meta 配置
 * @param params
 */
function meta(params) {
    var _a, _b;
    var chart = params.chart, options = params.options;
    var meta = options.meta, xAxis = options.xAxis, yAxis = options.yAxis, xField = options.xField, yField = options.yField, outliersField = options.outliersField;
    var yFieldName = Array.isArray(yField) ? constant_2.BOX_RANGE : yField;
    var baseMeta = {};
    // make yField and outliersField share y mate
    if (outliersField) {
        var syncName = constant_2.BOX_SYNC_NAME;
        baseMeta = (_a = {},
            _a[outliersField] = { sync: syncName, nice: true },
            _a[yFieldName] = { sync: syncName, nice: true },
            _a);
    }
    var scales = utils_2.deepAssign(baseMeta, meta, (_b = {},
        _b[xField] = utils_2.pick(xAxis, constant_1.AXIS_META_CONFIG_KEYS),
        _b[yFieldName] = utils_2.pick(yAxis, constant_1.AXIS_META_CONFIG_KEYS),
        _b));
    chart.scale(scales);
    return params;
}
/**
 * axis 配置
 * @param params
 */
function axis(params) {
    var chart = params.chart, options = params.options;
    var xAxis = options.xAxis, yAxis = options.yAxis, xField = options.xField, yField = options.yField;
    var yFieldName = Array.isArray(yField) ? constant_2.BOX_RANGE : yField;
    // 为 false 则是不显示轴
    if (xAxis === false) {
        chart.axis(xField, false);
    }
    else {
        chart.axis(xField, xAxis);
    }
    if (yAxis === false) {
        chart.axis(constant_2.BOX_RANGE, false);
    }
    else {
        chart.axis(yFieldName, yAxis);
    }
    return params;
}
/**
 * legend 配置
 * @param params
 */
function legend(params) {
    var chart = params.chart, options = params.options;
    var legend = options.legend, groupField = options.groupField;
    if (groupField) {
        if (legend) {
            chart.legend(groupField, legend);
        }
        else {
            // Grouped Box Chart default has legend, and it's position is `bottom`
            chart.legend(groupField, { position: 'bottom' });
        }
    }
    else {
        chart.legend(false);
    }
    return params;
}
exports.legend = legend;
/**
 * 样式
 * @param params
 */
function style(params) {
    var chart = params.chart, options = params.options;
    var xField = options.xField, yField = options.yField, boxStyle = options.boxStyle;
    var geometry = utils_1.findGeometry(chart, 'schema');
    var yFieldName = Array.isArray(yField) ? constant_2.BOX_RANGE : yField;
    /**
     * style 的几种情况
     * g.style({ fill: 'red' });
     * g.style('x*y*color', (x, y, color) => ({ fill: 'red' }));
     */
    if (util_1.isFunction(boxStyle)) {
        geometry.style(xField + "*" + yFieldName, function (_x, _y) {
            var _a;
            return boxStyle((_a = {},
                _a[xField] = _x,
                _a[yFieldName] = _y,
                _a));
        });
    }
    else if (util_1.isObject(boxStyle)) {
        geometry.style(boxStyle);
    }
    return params;
}
/**
 * tooltip 配置
 * @param params
 */
function tooltip(params) {
    var chart = params.chart, options = params.options;
    var tooltip = options.tooltip;
    if (tooltip !== undefined) {
        chart.tooltip(tooltip);
    }
    return params;
}
exports.tooltip = tooltip;
/**
 * 箱型图适配器
 * @param params
 */
function adaptor(params) {
    return utils_2.flow(field, outliersPoint, meta, axis, style, legend, tooltip, common_1.interaction, common_1.animation, common_1.theme)(params);
}
exports.adaptor = adaptor;
