// @antv/g2plot/esm/plots/treemap/interactions/actions/treemap-drill-down-action.js
import { __extends } from "tslib";
import { Action } from '@antv/g2';
import { get, isArray } from '@antv/util';
import { transformData } from '../../utils';
var TreemapDrillDownAction = /** @class */ (function (_super) {
    __extends(TreemapDrillDownAction, _super);
    function TreemapDrillDownAction() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        // 存储历史下钻 scale
        _this.cacheDataStack = [];
        return _this;
    }
    TreemapDrillDownAction.prototype.drill = function (data) {
        var view = this.context.view;
        var currentData = view.getData();
        var groupScales = view.getGroupScales();
        var hierarchyConfig = get(view, ['interactions', 'treemap-drill-down', 'cfg', 'hierarchyConfig'], {});
        // 重新 update 数据
        var drillData = transformData({
            data: data,
            colorField: get(groupScales, [0, 'field']),
            openDrillDown: true,
            hierarchyConfig: hierarchyConfig,
        });
        view.changeData(drillData);
        this.cacheDataStack.push(currentData);
    };
    TreemapDrillDownAction.prototype.click = function () {
        var data = get(this.context, ['event', 'data', 'data']);
        if (!data)
            return false;
        this.drill(data);
    };
    TreemapDrillDownAction.prototype.reset = function () {
        var view = this.context.view;
        if (!isArray(this.cacheDataStack) || this.cacheDataStack.length <= 0) {
            return;
        }
        var cacheData = this.cacheDataStack.splice(this.cacheDataStack.length - 1, 1);
        view.changeData(cacheData[0]);
    };
    return TreemapDrillDownAction;
}(Action));
export { TreemapDrillDownAction };
