// @antv/g2plot/lib/plots/sunburst/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sunburst = void 0;
var tslib_1 = require("tslib");
var utils_1 = require("../../utils");
var plot_1 = require("../../core/plot");
var adaptor_1 = require("./adaptor");
var Sunburst = /** @class */ (function (_super) {
    tslib_1.__extends(Sunburst, _super);
    function Sunburst() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'sunburst';
        return _this;
    }
    /**
     * 获取旭日图默认配置
     */
    Sunburst.prototype.getDefaultOptions = function () {
        return utils_1.deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            type: 'partition',
            innerRadius: 0,
            seriesField: 'value',
            tooltip: {
                shared: true,
                showMarkers: false,
                offset: 20,
                showTitle: false,
            },
        });
    };
    /**
     * 获取旭日图的适配器
     */
    Sunburst.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return Sunburst;
}(plot_1.Plot));
exports.Sunburst = Sunburst;
