// @antv/g2plot/lib/plots/heatmap/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Heatmap = void 0;
var tslib_1 = require("tslib");
var plot_1 = require("../../core/plot");
var utils_1 = require("../../utils");
var adaptor_1 = require("./adaptor");
// registered shapes
require("./shapes/circle");
require("./shapes/square");
var Heatmap = /** @class */ (function (_super) {
    tslib_1.__extends(Heatmap, _super);
    function Heatmap() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'heatmap';
        return _this;
    }
    /**
     * 获取直方图的适配器
     */
    Heatmap.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    Heatmap.prototype.getDefaultOptions = function () {
        return utils_1.deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            type: 'polygon',
            legend: false,
            xAxis: {
                tickLine: null,
                line: null,
                grid: {
                    alignTick: false,
                    line: {
                        style: {
                            lineWidth: 1,
                            lineDash: null,
                            stroke: '#f0f0f0',
                        },
                    },
                },
            },
            yAxis: {
                grid: {
                    alignTick: false,
                    line: {
                        style: {
                            lineWidth: 1,
                            lineDash: null,
                            stroke: '#f0f0f0',
                        },
                    },
                },
            },
        });
    };
    return Heatmap;
}(plot_1.Plot));
exports.Heatmap = Heatmap;
