// @antv/g2plot/lib/plots/treemap/interactions/treemap-drill-down.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var g2_1 = require("@antv/g2");
var treemap_drill_down_action_1 = require("./actions/treemap-drill-down-action");
var util_1 = require("./util");
g2_1.registerAction('treemap-drill-down-action', treemap_drill_down_action_1.TreemapDrillDownAction);
g2_1.registerInteraction('treemap-drill-down', {
    showEnable: [
        { trigger: 'plot:mouseenter', action: 'cursor:pointer' },
        { trigger: 'plot:mouseleave', action: 'cursor:default' },
    ],
    start: [
        {
            trigger: 'element:click',
            isEnable: util_1.isTopParentNode,
            action: ['treemap-drill-down-action:click', 'reset-button:show'],
        },
    ],
    rollback: [
        { trigger: 'reset-button:click', action: ['treemap-drill-down-action:reset'] },
        {
            trigger: 'reset-button:click',
            action: ['reset-button:hide'],
            isEnable: function (context) { return !util_1.hasHistoryDrill(context); },
        },
    ],
});
