// @antv/g2plot/lib/plots/radial-bar/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RadialBar = void 0;
var tslib_1 = require("tslib");
var plot_1 = require("../../core/plot");
var utils_1 = require("../../utils");
var adaptor_1 = require("./adaptor");
/**
 * 玉珏图
 */
var RadialBar = /** @class */ (function (_super) {
    tslib_1.__extends(RadialBar, _super);
    function RadialBar() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'radial-bar';
        return _this;
    }
    /**
     * @override
     * @param data
     */
    RadialBar.prototype.changeData = function (data) {
        this.updateOption({ data: data });
        // 更新玉珏图的scale
        adaptor_1.meta({ chart: this.chart, options: this.options });
        this.chart.changeData(data);
    };
    /**
     * 获取默认配置
     */
    RadialBar.prototype.getDefaultOptions = function () {
        return utils_1.deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            interactions: [{ type: 'element-active' }],
            legend: false,
            tooltip: {
                showMarkers: false,
            },
            xAxis: {
                grid: null,
                tickLine: null,
                line: null,
            },
            maxAngle: 240,
        });
    };
    /**
     * 获取适配器
     */
    RadialBar.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return RadialBar;
}(plot_1.Plot));
exports.RadialBar = RadialBar;
